process/track/baseBlockTrack.go

```diff
@@ -0,0 +1,425 @@
+
+	if shardID != bbt.shardCoordinator.SelfId() {
+		bbt.crossNotarizer.cleanupNotarizedHeadersBehindNonce(shardID, crossNotarizedNonce)
+		nonce = crossNotarizedNonce
+	}
+
+	bbt.cleanupTrackedHeadersBehindNonce(shardID, nonce)
+}
+
+func (bbt *baseBlockTrack) cleanupTrackedHeadersBehindNonce(shardID uint32, nonce uint64) {
+	if nonce == 0 {
+		return
+	}
+
+	bbt.mutHeaders.Lock()
+	defer bbt.mutHeaders.Unlock()
+
+	headersForShard, ok := bbt.headers[shardID]
+	if !ok {
+		return
+	}
+
+	for headersNonce := range headersForShard {
+		if headersNonce < nonce {
+			delete(headersForShard, headersNonce)
+		}
+	}
+}
+
+// ComputeLongestChain returns the longest valid chain for a given shard from a given header
+func (bbt *baseBlockTrack) ComputeLongestChain(shardID uint32, header data.HeaderHandler) ([]data.HeaderHandler, [][]byte) {
+	return bbt.blockProcessor.computeLongestChain(shardID, header)
+}
+
+// DisplayTrackedHeaders displays tracked headers
+func (bbt *baseBlockTrack) DisplayTrackedHeaders() {
+	for shardID := uint32(0); shardID < bbt.shardCoordinator.NumberOfShards(); shardID++ {
+		bbt.displayHeadersForShard(shardID)
+	}
+
+	bbt.displayHeadersForShard(sharding.MetachainShardId)
+}
+
+func (bbt *baseBlockTrack) displayHeadersForShard(shardID uint32) {
+	bbt.displayTrackedHeadersForShard(shardID, "tracked headers")
+	bbt.crossNotarizer.displayNotarizedHeaders(shardID, "cross notarized headers")
+	bbt.selfNotarizer.displayNotarizedHeaders(shardID, "self notarized headers")
+}
+
+func (bbt *baseBlockTrack) displayTrackedHeadersForShard(shardID uint32, message string) {
+	headers, hashes := bbt.sortHeadersFromNonce(shardID, 0)
+	shouldNotDisplay := len(headers) == 0 ||
+		len(headers) == 1 && headers[0].GetNonce() == 0
+	if shouldNotDisplay {
+		return
+	}
+
+	log.Debug(message,
+		"shard", shardID,
+		"nb", len(headers))
+
+	for index, header := range headers {
+		log.Trace("tracked header info",
+			"round", header.GetRound(),
+			"nonce", header.GetNonce(),
+			"hash", hashes[index])
+	}
+}
+
+// GetCrossNotarizedHeader returns a cross notarized header for a given shard with a given offset, behind last cross notarized header
+func (bbt *baseBlockTrack) GetCrossNotarizedHeader(shardID uint32, offset uint64) (data.HeaderHandler, []byte, error) {
+	return bbt.crossNotarizer.getNotarizedHeader(shardID, offset)
+}
+
+// GetLastCrossNotarizedHeader returns last cross notarized header for a given shard
+func (bbt *baseBlockTrack) GetLastCrossNotarizedHeader(shardID uint32) (data.HeaderHandler, []byte, error) {
+	return bbt.crossNotarizer.getLastNotarizedHeader(shardID)
+}
+
+// GetLastHeader returns the last header received (highest round) for the given shard
+func (bbt *baseBlockTrack) GetLastHeader(shardID uint32) data.HeaderHandler {
+	bbt.mutHeaders.RLock()
+	defer bbt.mutHeaders.RUnlock()
+
+	var lastHeaderForShard data.HeaderHandler
+
+	headersForShard, ok := bbt.headers[shardID]
+	if !ok {
+		return lastHeaderForShard
```

Comment:
this return is pretty problematic. as it in only a pointer, without the underlying structure. Those who use this function has to be pretty careful.

Reply:
Actully this method is used only in the exported method from this file -> IsShardSuck (which returns true or false). So I made it unexported